First cli iteration

sbjorn opened the pull request:
This is an implementation of the cli which will allow the Manila driver to manage nfs exports residing on the ring.

The base assumption is that this cli exists on a machine that can be accessible by openstack tenants, and the manila admin network, and has the nfs connector configured for access to an already installed and configured ring.

When access to an export for a client is granted or revoked, the exports file is modified accordingly and reexported by HUPing the sfused process.

Commit:
Add exports deserialization

Commit:
Add serialization of exports

Commit:
Outline export management interface

Commit:
Add context manager for privilege elevation

Commit:
Add environment checks

Commit:
Add support for creation of an export point

Commit:
Add support for exporting/unexporting a filesystem to a client

Commit:
Implement reload of export configuration

Write exports file and SIGHUP sfused.

Commit:
Add cli layer

Commit:
Unit test: Add test cases for Helper class

Commit:
Unit test: Add tests of utils

Commit:
Unit test: Add hypothesis strategy for generation of exports

NicolasT commented on scality_manila_utils/exceptions.py:
```diff
@@ -0,0 +1,26 @@
+# limitations under the License.
+
+
+class DeserializationException(Exception):
```
Would it make sense to let this one inherit from `ValueError`?

Also, aim for 100% docstring coverage.

NicolasT commented on scality_manila_utils/exceptions.py:
```diff
@@ -0,0 +1,26 @@
+    pass
+
+
+class EnvironmentException(Exception):
```
There's a builtin `EnvironmentError`, is this similar? Can it be reused instead of introducing a new exception type?

NicolasT commented on scality_manila_utils/export.py:
```diff
@@ -0,0 +1,219 @@
+        """
+        Create a new exports table from the given exports.
+
+        :param exports: a list, tuple or other iterable of exports
```
Just `iterable` will do :smile:

NicolasT commented on scality_manila_utils/export.py:
```diff
@@ -0,0 +1,219 @@
+            for export in exports
+        )
+
+    def add_client(self, export_point, host, options=None):
```
I'd change the name (and argument name) of this method. Exports files don't really handle 'clients', most of the time one uses a larger network (base IP + netmask) for a single export.

sbjorn replied:
This is the terminology used in the man page. I could change `host` to `machine_name` to make it closer to the man page terminology, but I don't think that's much better.

NicolasT commented on scality_manila_utils/export.py:
```diff
@@ -0,0 +1,219 @@
+        The export is created if it does not exist.
+
+        :param export_point: filesystem (path) to export
+        :type export_point: string
```
Maybe distinct between `unicode` and `bytes`? We're experiencing some issues where this distinction is unclear in `scality-sproxyd-client`, so better be clear from the beginning.

NicolasT commented on scality_manila_utils/export.py:
```diff
@@ -0,0 +1,219 @@
+            stripped = line.strip()
+            return stripped == '' or stripped.startswith('#')
+
+        return ExportTable(
```
As this is a `classmethod`, consider calling `cls(...)` here instead of `ExportTable`.

NicolasT replied:
`return cls(...`, or make it a `staticmethod`

JordanP replied:
Agreed. This should be a staticmethod.

NicolasT commented on scality_manila_utils/utils.py:
```diff
@@ -0,0 +1,123 @@
+    previous_gid = os.getegid()
+    # Become root
+    os.seteuid(0)
+    os.setegid(0)
```
This needs some changes to correctly handle errors. E.g., consider what happens when `os.setegid` fails? Same goes for the `finally` block.

Commit:
Implement exports comparison functions

Commit:
Unit test: Add tests of export (de)serialization

Commit:
Unit test: Add tests of add/removal of exports

NicolasT commented on manila-cli.py:
```diff
@@ -0,0 +1,110 @@
+#!/usr/bin/env python
```
Instead of using a top-level script like this, create a `scality_manila_utils.cli` module with a `main(args=None)` procedure, and add a `console_script` using `setuptools`. You can add and `if __name__ == '__main__': main()` in there to be able to launch it for testing purposes, then retrieve `sys.argv` if `args` is `None`, or use the provided args for parsing etc, which aids writing unit tests. See https://github.com/scality/ScalitySproxydSwift/commit/49512f4d09dfe0280cc8c9e4e23eec5a8bebf0ba for an example.

NicolasT commented on manila-cli.py:
```diff
@@ -0,0 +1,110 @@
+
+    # Attempt to find a proper user and group
+    user_info = get_user('nobody')
+    group_info = get_group('nogroup')
```
FWIW, my Fedora system has no `nogroup` group, it has `nobody`.

NicolasT replied:
Instead of doing an initial lookup and one fallback, I'd create a list of 'usual suspects', iterate over it and use the first available.

NicolasT commented on manila-cli.py:
```diff
@@ -0,0 +1,110 @@
+from scality_manila_utils.helper import Helper
+
+
+def drop_privileges():
```
Add a bit of logging throughout the scripts (including in this procedure).

NicolasT commented on manila-cli.py:
```diff
@@ -0,0 +1,110 @@
+        group_info = get_group('nobody')
+
+    # Drop privileges
+    if user_info and group_info:
```
So, if it can't find the expected user and/or group, it doesn't do anything. I'd raise instead of happily continuing.

sbjorn replied:
Thanks, that needs to be fixed.

NicolasT commented on manila-cli.py:
```diff
@@ -0,0 +1,110 @@
+    # Drop any elevated permissions
+    drop_privileges()
+
+    pre_parser = argparse.ArgumentParser(add_help=False)
```
Don't create the parser in a top-level section, this way it can't be tested.

NicolasT commented on manila-cli.py:
```diff
@@ -0,0 +1,110 @@
+    pre_parser = argparse.ArgumentParser(add_help=False)
+    pre_parser.add_argument('--exports', help='Path to exports file',
+                            default='/etc/exports.conf')
+    pre_parser.add_argument('--basedir', help='Directory where manila '
```
Is this required? The script will need to NFS-mount `localhost:/`, so it could create a temporary directory to do so (which removes potential concurrency issues).

NicolasT commented on manila-cli.py:
```diff
@@ -0,0 +1,110 @@
+                            'filesystems reside, and are reexported',
+                            default='/mnt/manila')
+
+    command_parser = argparse.ArgumentParser(parents=[pre_parser],
```
Layout: consider something like

``` python
command_parser = argparse.ArgumentParser(
    parents=[pre_parser, description='...')
```

which could remove the need for line-breaks in the description string etc.

NicolasT commented on manila-cli.py:
```diff
@@ -0,0 +1,110 @@
+    parser_check = subparsers.add_parser('check', help='Check the environment '
+                                         'for required binaries and running '
+                                         'services')
+
```
Is there a way to query the version of the tool, so the Manila driver can query this and validate in `check_configuration`?

NicolasT commented on scality_manila_utils/exceptions.py:
```diff
@@ -0,0 +1,26 @@
+    """Raised on errors pertaining to management of exports."""
+
+
+class EnvironmentException(Exception):
```
There's a builtin EnvironmentError, is this similar? Can it be reused instead of introducing a new exception type?

sbjorn replied:
The `EnvironmentError` seems to be for lower-level errors, closer to the OS.

NicolasT commented on scality_manila_utils/export.py:
```diff
@@ -0,0 +1,250 @@
+    def __ne__(self, other):
+        eq = self == other
+
+        if isinstance(eq, bool):
```
I believe the result of an `==` operation is always a bool, or an exception. The way to do it is calling `__eq__` explicitly (without the magic `==` puts around it), and act appropriately:

``` python
def __ne__(self, other):
    res = self.__eq__(other)
    if res is NotImplemented:
        return res
    else:
        return (not res)
```

NicolasT commented on scality_manila_utils/export.py:
```diff
@@ -0,0 +1,250 @@
+            return not eq
+        else:
+            return NotImplemented
+
```
Consider implementing `__repr__` and `__hash__` for well-behaving data objects.

sbjorn replied:
`__repr__` is nice to have, and will be added. I don't think `__hash__` makes much sense for the ExportTable, as it is mutable.

NicolasT commented on scality_manila_utils/export.py:
```diff
@@ -0,0 +1,250 @@
+            host = match.group('host')
+            option_list = match.group('options')
+            if option_list is None:
+                options = set()
```
Can this be a `frozenset`? (Same below)

sbjorn replied:
Makes total sense.

NicolasT commented on scality_manila_utils/export.py:
```diff
@@ -0,0 +1,250 @@
+            if options:
+                clients += '({0:s})'.format(','.join(options))
+
+        export_line = '{export_point:<32s} {clients:s}'.format(
```
Can you provide a reference for the 32 characters restriction?

sbjorn replied:
This is just an attempt to have some alignment of the clients in the exports file. There is nothing more to it.

JordanP replied:
Maybe add a code comment for future reference ? (it's just one line... :))

NicolasT commented on scality_manila_utils/utils.py:
```diff
@@ -0,0 +1,123 @@
+                                   "started".format(process))
+
+
+def safe_write(text, path, permissions=0o755):
```
I don't think this will work as expected. `rename` doesn't work cross-filesystem: `tempfile` will create a file in some temporary directory (which, when `/tmp`, could be vulnerable to a symlink attack if I'm not mistaken), and `rename` would fail if your target is e.g. `/etc/foo` and `/tmp` is some `tmpfs`.

Commit:
Implement the `__repr__` methods of the export classes

Commit:
Add logging

Commit:
Unit test: Add test of cli layer

Commit:
Tox configuration: Don't flake hypothesis working directory

Commit:
Add version

NicolasT commented on scality_manila_utils/cli.py:
```diff
@@ -0,0 +1,190 @@
+    log_format = logging.Formatter(
+        '%(asctime)s - %(name)s - %(levelname)s - %(message)s'
+    )
+    handler = logging.handlers.SysLogHandler('/dev/log')
```
Instead of logging to syslog on the NFS server system, consider (also) logging to `stderr` so the calling system (`manila-api`) can capture logging and log the output to its own log infrastructure? This provides (useful) information to an admin when debugging Manila issues.

sbjorn replied:
Any exceptions raised, and an accompanying stack trace will be available on the manila side, as it is not supressed, but bubbled up to the command line. I'm not sure that it adds much extra value esssentially log everything over ssh.

NicolasT commented on scality_manila_utils/cli.py:
```diff
@@ -0,0 +1,190 @@
+        log.debug('Dropping privileges to %s:%s', user_info.pw_name,
+                  group_info.gr_name)
+        os.setegid(group_info.gr_gid)
+        os.seteuid(user_info.pw_uid)
```
Reset `egid` when this fails (for whatever reason).

sbjorn replied:
An `OSError` will be thrown, and the program will halt when this happens.

JordanP replied:
What if the `OSError` is caught at a higher level ?  would that make the program to continue with an unexpected egid ? Note that I have no idea about what I am talking about... just asking based on the code and comments I am reading.

NicolasT commented on scality_manila_utils/cli.py:
```diff
@@ -0,0 +1,190 @@
+def main(args=None):
+    setup_logger()
+
+    if os.geteuid() != 0:
```
I believe this should be `getuid`: if `euid` equals `0`, it'd be impossible to raise privileges back to `0` once dropped.

NicolasT commented on scality_manila_utils/cli.py:
```diff
@@ -0,0 +1,190 @@
+    pre_parser.add_argument(
+        '--exports',
+        help='Path to exports file',
+        default='/etc/exports.conf'
```
Shouldn't this be `/etc/exports`?

sbjorn replied:
No, the sfused NFS connector is looking for `/etc/exports.conf`.

NicolasT commented on scality_manila_utils/cli.py:
```diff
@@ -0,0 +1,190 @@
+        log.error('Invoked without superuser privileges')
+        raise RuntimeError("This program requires superuser privileges")
+
+    pre_parser = argparse.ArgumentParser(add_help=False)
```
Add a `--version` thingy so the Manila driver can query the version of `scality-manila-utils` installed on a remote system, and validate it's a supported version during it's system/configuration checks.

sbjorn replied:
Good point, I will add that.

NicolasT commented on scality_manila_utils/cli.py:
```diff
@@ -0,0 +1,190 @@
+        logging.root.setLevel(logging.DEBUG)
+
+    # Drop any elevated permissions
+    drop_privileges()
```
I think this could be done earlier within `main`?

sbjorn replied:
I considered that as well, but logging must be setup before that call. Otherwise logs pertaining to `drop_privileges` will be lost. It is a small tradeoff.

NicolasT commented on scality_manila_utils/cli.py:
```diff
@@ -0,0 +1,190 @@
+        for arg, val in command_args.items()
+    )
+    log.info("Invoking %s(%s)", parsed_args.func.__name__, formatted_args)
+    try:
```
Instead of having a couple of `try`/`except` clauses inside `main` which then `log.exception` and `raise`, I'd wrap `main` in a decorator which does this for any (unhandled) exception.

NicolasT commented on scality_manila_utils/export.py:
```diff
@@ -0,0 +1,264 @@
+        :param options: sequence of nfs options (optional)
+        :type options: iterable of strings
+        """
+        if options is None:
```
Consider `export_options = frozenset(options or [])`

JordanP replied:
This has been addressed (see next line)

NicolasT commented on scality_manila_utils/export.py:
```diff
@@ -0,0 +1,264 @@
+        log.debug("'%s' revoked from export '%s'", host, export_point)
+
+    @classmethod
+    def deserialize(cls, export_content):
```
Would it be possible to keep any comments in the file and re-insert them when writing out the file?

sbjorn replied:
It would certainly be possible, but I don't think it is needed for a first version.

NicolasT commented on scality_manila_utils/export.py:
```diff
@@ -0,0 +1,264 @@
+
+    def __repr__(self):
+        exports = (repr(export) for export in self.exports.values())
+        return "ExportTable({0:s})".format(", ".join(exports))
```
It's nice to have `__repr__` return a string which can be copy-pasted into a REPL to construct the same object. I guess adding square brackets in the constructor call here should do. Furthermore, consider using named arguments.

NicolasT commented on scality_manila_utils/export.py:
```diff
@@ -0,0 +1,264 @@
+
+        export_point = export_parts[0]
+        clients = dict(map(client_extract, export_parts[1:]))
+        return Export(export_point, clients)
```
Consider `return cls(...`

NicolasT commented on scality_manila_utils/export.py:
```diff
@@ -0,0 +1,264 @@
+    def __ne__(self, other):
+        eq = self.__eq__(other)
+
+        if isinstance(eq, bool):
```
Any reason why to prefer an instance check over `is` equality to `NotImplemented`?

NicolasT commented on scality_manila_utils/helper.py:
```diff
@@ -0,0 +1,146 @@
+        self.root_volume = root_volume
+
+        if not os.path.exists(exports_path):
+            raise EnvironmentException('Unable to locate exports file')
```
Is an existing `/etc/exports` file required?

sbjorn replied:
Not strictly required, but `/etc/exports.conf` is generated when the nfs connector and volume is setup. It should probably not be absent. We could log a warning if that is more appropriate.

NicolasT commented on scality_manila_utils/helper.py:
```diff
@@ -0,0 +1,146 @@
+        self.exports_path = exports_path
+
+        with io.open(self.exports_path, 'rt') as exports_file:
+            lines = (line for line in exports_file)
```
So, `lines` = `exports_file` from a behaviour PoV? Then I guess `exports_file` can be passed to `ExportTable.deserialize` as-is :-)

sbjorn replied:
Doh! :-)

NicolasT commented on scality_manila_utils/helper.py:
```diff
@@ -0,0 +1,146 @@
+        with utils.elevated_privileges():
+            utils.safe_write(exports_data, self.exports_path)
+            for pid in sfused_pids:
+                os.kill(pid, signal.SIGHUP)
```
Consider adding some logging here.

NicolasT commented on scality_manila_utils/helper.py:
```diff
@@ -0,0 +1,146 @@
+
+        :param export_name: name of export to remove
+        :type export_name: string (unicode)
+        """
```
`raise NotImplementedError`?

NicolasT commented on scality_manila_utils/helper.py:
```diff
@@ -0,0 +1,146 @@
+
+        :param export_name: name of export
+        :type export_name: string (unicode)
+        """
```
`raise NotImplementedError`?

NicolasT commented on scality_manila_utils/helper.py:
```diff
@@ -0,0 +1,146 @@
+                                             ExportException)
+
+
+class Helper(object):
```
I'm a bit dubious on the use of a class for this, and making is stateful (`self.exports`) for CLI usage. Any benefit over plain procedures in a module which read/update/write the exports file (or whatever actions they ought to take) with all required parameters passed as arguments?

NicolasT commented on scality_manila_utils/utils.py:
```diff
@@ -0,0 +1,177 @@
+    # Become root
+    log.debug("Elevating privileges")
+    os.seteuid(0)
+    os.setegid(0)
```
This could (in theory, but anyway) fail, and needs to reset euid before raising.

NicolasT commented on scality_manila_utils/utils.py:
```diff
@@ -0,0 +1,177 @@
+    finally:
+        # Drop root privileges
+        log.debug("Dropping elevated privileges")
+        os.setegid(previous_gid)
```
(Try to) drop euid even if this fails.

NicolasT commented on scality_manila_utils/utils.py:
```diff
@@ -0,0 +1,177 @@
+            if e.errno != errno.ENOENT:
+                raise
+
+    log.debug("PIDs for '%s': %s", process, process_pids)
```
Strictly speaking, `process_pids` is a list, so `%r` to format it would be preferable (although for `list` indeed `str` and `repr` are the same).

NicolasT commented on scality_manila_utils/utils.py:
```diff
@@ -0,0 +1,177 @@
+    return process_pids
+
+
+def binary_check(binary, paths):
```
Nothing wrong with this procedure, though I'd let it return a `bool` and let whoever is calling handle the failure case appropriately, without requiring exception handling.

NicolasT commented on scality_manila_utils/utils.py:
```diff
@@ -0,0 +1,177 @@
+    with tempfile.NamedTemporaryFile(mode='wt', dir=target_dir,
+                                     delete=False) as f:
+        os.chmod(f.name, permissions)
+        f.write(text)
```
Call `f.flush()` before syncing.

JordanP replied:
What was the outcome of this comment ?

JordanP replied:
This has been addressed (see next line)

NicolasT commented on scality_manila_utils/utils.py:
```diff
@@ -0,0 +1,177 @@
+            subprocess.check_call(['umount', mount_point])
+            log.debug('Unmounted nfs root')
+            os.rmdir(mount_point)
+        except subprocess.CalledProcessError:
```
Consider moving the `except` clauses more closely to their origin: one for the `umount` call, one for the `rmdir` call, both only handling the expected exceptions.

JordanP replied:
+1 (ie 2 try/except blocks)

NicolasT commented on setup.cfg:
```diff
@@ -26,6 +26,10 @@ classifier =
 
+[entry_points]
+console_scripts =
+    manila-cli = scality_manila_utils.cli:main
```
Better call this `scality-manila-utils` or something alike, `manila-cli` could be from the upstream Manila project.

NicolasT commented on test-requirements.txt:
```diff
@@ -1,2 +1,5 @@
 coverage
+hypothesis
+mock
+unittest2
```
Out of interest, any reason to use `unittest2`?

sbjorn replied:
It has a bunch of assertion methods backported from 2.7

Commit:
Implement `get_export`

Commit:
Introduce more fine-grained exceptions where appropriate

Commit:
Get exit code from exception on error path

Commit:
Mark `wipe_export` as unimplemented

Commit:
Make sure that created exports are world read-writable

JordanP commented on scality_manila_utils/helper.py:
```diff
@@ -0,0 +1,180 @@
+                    os.mkdir(export_point)
+                    os.chmod(export_point, 0o0777)
+
+    def remove_export(self, export_name):
```
Suggestion: this could be renamed to "wipe_export'. Or update the doc string to say this method should actually delete data.

JordanP commented on scality_manila_utils/export.py:
```diff
@@ -0,0 +1,273 @@
+
+class ExportTable(object):
+    """
+    A set of exports that can be destilled into /etc/exports.
```
distilled

JordanP commented:
You should add mock and unittest2 in test-requirements.txt

Commit:
Add a pylint Tox target

Also, do not explicitely install test-requirements.txt, it's not needed
because it's inherited from the generic [testenv] section.

Commit:
Add a "main" entry point and print default values when --help is requested

Commit:
Prefer list comprehension over filter+lambda

Commit:
Don't print timestamp to logs

A timestamp is usually attached by the logging facility.

JordanP commented:
LGTM
